rust/src/transformations/make_stable_expr/expr_discrete_quantile_score/plugin_dq_score.rs

```diff
@@ -0,0 +1,209 @@
+            collect_groups: ApplyOptions::GroupWise,
+            fmt_str: "dq_score",
+            returns_scalar: true,
+            changes_length: true,
+            ..Default::default()
+        }
+    }
+}
+
+// allow the LaplaceArgs struct to be stored inside an AnonymousFunction, when used from Rust directly
+impl SeriesUdf for DQScoreArgs {
+    // makes it possible to downcast the AnonymousFunction trait object back to LaplaceArgs
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn call_udf(&self, s: &mut [Series]) -> PolarsResult<Option<Series>> {
+        dq_score_udf(s, self.clone()).map(Some)
+    }
+
+    fn get_output(&self) -> Option<GetOutput> {
+        let kwargs = self.clone();
+        Some(GetOutput::map_fields(move |fields| {
+            dq_score_type_udf(fields, kwargs.clone())
+                .ok()
+                .unwrap_or_else(|| fields[0].clone())
+        }))
+    }
+}
+
+/// Implementation of the Laplace noise expression.
+///
+/// The Polars engine executes this function over chunks of data.
+fn dq_score_udf(inputs: &[Series], kwargs: DQScoreArgs) -> PolarsResult<Series> {
+    let Ok([series]) = <&[_; 1]>::try_from(inputs) else {
+        polars_bail!(InvalidOperation: "Quantile expects a single input field");
+    };
+
+    let n = series.len() as u64;
+    let (candidates, constants) = (kwargs.candidates, kwargs.constants);
+    let Some((alpha_num, alpha_den, size_limit)) = constants else {
+        polars_bail!(InvalidOperation:
+            "encountered quantile in expression that has not been made private or stable",
+        );
+    };
+
+    // PT stands for Polars Type
+    fn hist<PT: 'static + PolarsDataType>(
+        series: &Series,
+        candidates: Vec<f64>,
+    ) -> PolarsResult<(Vec<u64>, Vec<u64>)>
+    where
+        // candidates must be able to be converted into a the physical dtype
+        for<'a> PT::Physical<'a>: RoundCast<f64> + PartialOrd,
+        PT::Array: StaticArray,
+    {
+        let candidates = candidates
+            .into_iter()
+            .map(PT::Physical::round_cast)
+            .collect::<Fallible<Vec<_>>>()?;
+
+        // count of the number of records between...
+        //  (-inf, c1), [c1, c2), [c2, c3), ..., [ck, inf)
+        let mut hist_lo = vec![0u64; candidates.len() + 1];
+        //  (-inf, c1], (c1, c2], (c2, c3], ..., (ck, inf)
+        let mut hist_ro = vec![0u64; candidates.len() + 1];
+
+        series
+            .unpack::<PT>()?
+            .downcast_iter()
+            .flat_map(StaticArray::values_iter)
+            .for_each(|v| {
+                let idx_lt = candidates.partition_point(|c| *c < v);
+                hist_ro[idx_lt] += 1;
+
+                let idx_eq = idx_lt + candidates[idx_lt..].partition_point(|c| *c == v);
+                hist_lo[idx_eq] += 1;
+            });
+
+        // don't care about the number of elements greater than all
+        hist_lo.pop();
+        hist_ro.pop();
```

Comment:
Add a test of just `dq_score_udf` or even pull this inner part into its own function? Seems good, but also the kind of place where it'd be easy to overlook and off-by-one error.